From 38385c0a9503a443a3bbaab45e8893bcd807295e Mon Sep 17 00:00:00 2001
From: Lanttor Guo <lanttor.guo@freescale.com>
Date: Wed, 9 Jun 2010 10:41:44 +0800
Subject: [PATCH] Fix Bug of USB OTG auto-detect issue on m5441x platform

This because hardware collision between FEC2 and ULPI.
Work around to run USB OTG when disableing FEC2 clock;

Another fix is to enable USB OTG external pulldown resistor for HOST.

Signed-off-by: Lanttor Guo <lanttor.guo@freescale.com>
---
 arch/m68k/coldfire/m5441x/usb.c  |    3 ---
 arch/m68k/coldfire/m5441x/xcvr.c |   23 +++++++++++++++++++----
 2 files changed, 19 insertions(+), 7 deletions(-)

diff --git a/arch/m68k/coldfire/m5441x/usb.c b/arch/m68k/coldfire/m5441x/usb.c
index bfe9b6f..aebca5c 100644
--- a/arch/m68k/coldfire/m5441x/usb.c
+++ b/arch/m68k/coldfire/m5441x/usb.c
@@ -43,9 +43,6 @@ static inline void fsl_usb_enable_clk(void)
 {
 	pr_debug("%s\n", __func__);
 
-        /* Enable the clock for USB OTG */
-        MCF_PM_PPMCR0 = 0x2c;
-
 	/* Use external clock source if PLL isn't a multiple of 60MHz */
 	MCF_CCM_MISCCR &= ~MCF_CCM_MISCCR_USBSRC;
 
diff --git a/arch/m68k/coldfire/m5441x/xcvr.c b/arch/m68k/coldfire/m5441x/xcvr.c
index 79a20d9..8d52f44 100644
--- a/arch/m68k/coldfire/m5441x/xcvr.c
+++ b/arch/m68k/coldfire/m5441x/xcvr.c
@@ -13,6 +13,7 @@
 #include <linux/types.h>
 #include <linux/init.h>
 #include <linux/io.h>
+#include <linux/delay.h>
 #include <linux/usb/fsl_xcvr.h>
 #include <asm/mcfsim.h>
 
@@ -34,9 +35,11 @@ static void xcvr_fsls_init(struct fsl_xcvr_ops *this)
 {
 	pr_debug("%s: name=%s\n", __func__, this->name);
 
-	/* Configure USB_DP_PDOWN/USB_DM_PDOWN */
+	/* Configure USB_DP_PDOWN/USB_DM_PDOWN (external resistor) */
 	MCF_GPIO_PAR_DBGH1 &= ~MCF_GPIO_PAR_DEBUGH1_DAT0;
 	MCF_GPIO_PAR_DBGH1 &= ~MCF_GPIO_PAR_DEBUGH1_DAT1;
+	MCF_GPIO_PDDR_I |= 0xc0;
+	MCF_GPIO_PODR_I &= 0x3f;
 
 	/* Enable VBUS_EN and VBUS_OC signals */
 	MCF_GPIO_PAR_TIMER =
@@ -49,9 +52,6 @@ static void xcvr_fsls_init(struct fsl_xcvr_ops *this)
 		MCF_GPIO_PAR_TIMER_T0IN_MASK) |
 		MCF_GPIO_PAR_TIMER_T0IN_USBO_VBUSOC;
 
-	/* Enable on-chip transceiver pull-down */
-	MCF_CCM_UOCSR |= MCF_CCM_UOCSR_XPDE;
-
 	/* Setup USB_VBUS_OC signal to be active-low */
 	MCF_CCM_MISCCR &= (~MCF_CCM_MISCCR_USBOOC);
 	MCF_CCM_MISCCR |= MCF_CCM_MISCCR_USBPUE;
@@ -65,6 +65,21 @@ static void xcvr_ulpi_init(struct fsl_xcvr_ops *this)
 {
 	pr_debug("%s: name=%s\n", __func__, this->name);
 
+	/* disable FEC2 clock */
+	MCF_PM_PPMSR0 = 54;
+
+	/* reset USB3300 */
+	asm ("moveb #0x40, %%d0\n\t"
+	     "moveb %%d0, 0xec090000\n\t"
+	     : : : "%d0");
+	mdelay(1);
+	asm ("moveb #0x0, %%d0\n\t"
+	     "moveb %%d0, 0xec090000\n\t"
+	     : : : "%d0");
+
+	/* enable USB OTG clock*/
+	MCF_PM_PPMCR0 = 44;
+
 	/* Chip select for ULPI: GPIO G4 as CS between ULPI and RMII1*/
 	MCF_GPIO_PAR_SIMP0H =
 		(MCF_GPIO_PAR_SIMP0H &
-- 
1.6.3.2

